src/service/video-manager-impl.js

```diff
@@ -0,0 +1,310 @@
+
+    /** @const @private {!../../src/service/vsync-impl.Vsync} */
+    this.vsync_ = vsyncFor(win);
+
+    /** @private {boolean} */
+    this.canAutoplay_ = video.hasAutoplay() &&
+        platformSupportsAutoplay(platformFor(win));
+
+    const element = dev().assert(video.element);
+
+    listenOncePromise(element, VideoEvents.BUILT)
+      .then(() => this.videoBuilt_());
+
+    listenOncePromise(element, VideoEvents.CAN_PLAY)
+      .then(() => this.videoLoaded_());
+  }
+
+  /**
+   * Called when the video element is built.
+   * @private
+   */
+  videoBuilt_() {
+    if (this.canAutoplay_) {
+      this.autoplayVideoBuilt_();
+    }
+  }
+
+  /**
+   * Called when the video is loaded and can play.
+   * @private
+   */
+  videoLoaded_() {
+    this.loaded_ = true;
+    if (this.isVisible_) {
+      // Handles the case when the video becomes visible before loading
+      this.loadedVideoVisibilityChanged_();
+    }
+  }
+
+  /**
+   * Called when visibility of a video changes.
+   * @private
+   */
+  videoVisibilityChanged_() {
+    if (this.loaded_) {
+      this.loadedVideoVisibilityChanged_();
+    }
+  }
+
+  /**
+   * Only called when visibility of a loaded video changes.
+   * @private
+   */
+  loadedVideoVisibilityChanged_() {
+    if (this.canAutoplay_) {
+      this.autoplayLoadedVideoVisibilityChanged_();
+    }
+  }
+
+  /**
+   * Whether there is a need to monitor visibility of this video.
+   * @return {boolean}
+   * @package
+   */
+  needsVisibilityObserver() {
+    return this.canAutoplay_;
+  }
+
+  /* Autoplay Behaviour */
+
+  /**
+   * Called when an autoplay video is built.
+   * @private
+   */
+  autoplayVideoBuilt_() {
+    this.video.mute();
+
+    // If autoplay video has controls, hide them and only show them on
+    // user-ineraction.
+    if (this.video.hasControls()) {
+      this.video.hideControls();
+
+      // TODO(aghassemi): This won't work for iframes, needs a transparent shim
+      listenOnce(this.video.element, 'click', () => {
+        this.video.showControls();
+        this.video.unmute();
+      });
+    }
+  }
+
+  /**
+   * Called when visibility of a loaded autoplay video changes.
+   * @private
+   */
+  autoplayLoadedVideoVisibilityChanged_() {
+    if (this.isVisible_) {
+      this.video.play(/*autoplay*/ true);
+    } else {
+      this.video.pause();
+    }
+  }
+
+  /**
+   * Called by all possible events that might change the visibility of the video
+   * such as scrolling or {@link ../video-interface.VideoEvents#VISIBILITY}.
+   * @package
+   */
+  updateVisibility() {
+    const wasVisible = this.isVisible_;
+
+    // Measure if video is now in viewport and what percentage of it is visible
+    const measure = () => {
+      if (!this.video.isInViewport()) {
+        this.isVisible_ = false;
+        return;
+      }
+
+      // Calculate what percentage of the video is in viewport
+      const change = this.video.element.getIntersectionChangeEntry();
```

Comment:
Thinking about it, it is probably still good to have it, if we need the
mutate block, as that makes sure the measurements are the latest at the
time of mutate.

On Wed, Aug 31, 2016 at 2:29 PM, Dima Voytenko <notifications@github.com>
wrote:

> In src/service/video-manager-impl.js
> <https://github.com/ampproject/amphtml/pull/4786#discussion_r77077194>:
>
> > +   * Called by all possible events that might change the visibility of the video
> > +   * such as scrolling or {@link ../video-interface.VideoEvents#VISIBILITY}.
> > +   * @package
> > +   */
> > +  updateVisibility() {
> > +    const wasVisible = this.isVisible_;
> > +
> > +    // Measure if video is now in viewport and what percentage of it is visible
> > +    const measure = () => {
> > +      if (!this.video.isInViewport()) {
> > +        this.isVisible_ = false;
> > +        return;
> > +      }
> > +
> > +      // Calculate what percentage of the video is in viewport
> > +      const change = this.video.element.getIntersectionChangeEntry();
>
> Correct. Measure block is not needed. These are always cached.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ampproject/amphtml/pull/4786/files/5e55b6681c220ac6c11372aa688dc175547184a0#r77077194>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFeT5vAmvPhIXXyRQfFHg8H3D1aBzmaks5qlfJDgaJpZM4JxI8M>
> .
>